### PR TITLE
SLE-373 show effective rule param value in the rule description

### PR DESCRIPTION
**Pull request by kirill-knize-sonarsource**

### CODE REVIEW

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintWebView.java [hunk 1]
```diff
@@ -85,7 +90,8 @@ public SonarLintWebView(Composite parent, boolean useEditorFontSize) {
       browser = new Browser(this, SWT.FILL);
       addLinkListener(browser);
       browser.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, true, 1, 1));
-      browser.setJavascriptEnabled(false);
+      browser.setJavascriptEnabled(true);
```
Comment (henryju) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintWebView.java:93]: As discussed, I now think we should move away from this `BrowserFunction` + JavaScript solution. I tried with a simple addition to the `LocationListener`, and it works.

**Review by henryju — CHANGES_REQUESTED** [review]
I think the HTML layout could be improved.
image

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintWebView.java [hunk 2]
```diff
@@ -199,10 +189,14 @@ public void changing(LocationEvent event) {
 
         event.doit = false;
 
-        try {
-          PlatformUI.getWorkbench().getBrowserSupport().getExternalBrowser().openURL(new URL(loc));
-        } catch (PartInitException | MalformedURLException e) {
-          SonarLintLogger.get().error("Unable to open URL: " + loc, e);
+        if(RULES_CONFIGURATION_LINK.equals(loc)) {
```
Comment (henryju) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintWebView.java:192]: formatting issue

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintWebView.java [hunk 3]
```diff
@@ -199,10 +189,14 @@ public void changing(LocationEvent event) {
...
+        	openSonarLintPreferences();
+        } else {
+        	try {
```
Comment (henryju) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintWebView.java:195]: indentation look wrong

**Review by henryju — CHANGES_REQUESTED** [review]

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintRuleBrowser.java [hunk 4]
```diff
@@ -67,9 +78,61 @@ protected String body() {
         + "<img class=\"severityicon\" alt=\"" + severity + "\" src=\"data:image/gif;base64," + severityImg64 + "\">"
         + "<span>" + clean(severity) + "</span>"
         + "</div>"
-        + "<div class=\"rule-desc\">" + htmlDescription
-        + "</div>";
+        + "<div class=\"rule-desc\">" + htmlDescription + "</div>"
+        + "<div>" + ruleDetailsMarkup + "</div>";
```
[the remark this answers is not in the document]
Comment (kirill-knize-sonarsource) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintRuleBrowser.java]: Done.

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintRuleBrowser.java [hunk 5]
```diff
@@ -55,6 +62,10 @@ protected String body() {
           htmlDescription += "<div>" + extendedDescription + "</div>";
         }
       }
+      String ruleDetailsMarkup = "";
+      if(ruleDetails instanceof StandaloneRuleDetails) {
```
[the remark this answers is not in the document]
Comment (kirill-knize-sonarsource) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintRuleBrowser.java]: Done.

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintRuleBrowser.java [hunk 6]
```diff
@@ -67,9 +78,61 @@ protected String body() {
...
+    }
+  }
+
+
+  private String renderRuleParams(StandaloneRuleDetails ruleDetails) {
+    if (!ruleDetails.paramDetails().isEmpty()) {
+      return "<script language=\"JavaScript\">\n" +
+        "function openSonarLintSettings() {\n" +
+        "openSonarLintPreferences();" +
+        //"alert('Alert from openSonarLintSettings function!');" +
```
[the remark this answers is not in the document]
Comment (kirill-knize-sonarsource) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintRuleBrowser.java]: Done.

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintRuleBrowser.java [hunk 7]
```diff
@@ -67,9 +78,61 @@ protected String body() {
...
+        + "<div class=\"rule-desc\">" + htmlDescription + "</div>"
+        + "<div>" + ruleDetailsMarkup + "</div>";
+    }
+  }
+
+
+  private String renderRuleParams(StandaloneRuleDetails ruleDetails) {
+    if (!ruleDetails.paramDetails().isEmpty()) {
+      return "<script language=\"JavaScript\">\n" +
+        "function openSonarLintSettings() {\n" +
```
[the remark this answers is not in the document]
Comment (kirill-knize-sonarsource) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintRuleBrowser.java]: Implementation changed to link listener.

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintRuleBrowser.java [hunk 8]
```diff
@@ -67,9 +78,61 @@ protected String body() {
...
+    }
+  }
+
+  private static String renderRuleParam(StandaloneRuleParam param, StandaloneRuleDetails ruleDetails) {
+    String paramDescription = param.description() != null ? "<p>" + param.description() + "</p>" : "";
+    String paramDefaultValue = param.defaultValue();
+    String defaultValue = paramDefaultValue != null ? paramDefaultValue : "(none)";
+    String currentValue = getRuleParamValue(ruleDetails.getKey(), param.name()).orElse(defaultValue);
+    return "<tr class='tbody'>" +
+      // The <br/> elements are added to simulate a "vertical-align: top" (not supported by Java 11 CSS renderer)
```
[the remark this answers is not in the document]
Comment (kirill-knize-sonarsource) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintRuleBrowser.java]: Removed.

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintWebView.java [hunk 9]
```diff
@@ -109,6 +115,23 @@ public SonarLintWebView(Composite parent, boolean useEditorFontSize) {
     refresh();
   }
 
+  private static void openSonarLintPreferences(Shell shell) {
+    PreferenceDialog dialog = PreferencesUtil.createPreferenceDialogOn(
+      shell, "org.sonarlint.eclipse.ui.properties.RulesConfigurationPage",
```
[the remark this answers is not in the document]
Comment (kirill-knize-sonarsource) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintWebView.java]: Done.

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintWebView.java (hunk 1 above)
Comment (kirill-knize-sonarsource) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintWebView.java]: Done.

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintWebView.java (hunk 2 above)
Comment (kirill-knize-sonarsource) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintWebView.java]: Fixed.

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintWebView.java (hunk 3 above)
Comment (kirill-knize-sonarsource) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintWebView.java]: Fixed.

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintWebView.java [hunk 10]
```diff
@@ -47,15 +48,19 @@
...
+import org.eclipse.ui.dialogs.PreferencesUtil;
 import org.sonarlint.eclipse.core.SonarLintLogger;
 
 import static org.eclipse.jface.preference.JFacePreferences.INFORMATION_BACKGROUND_COLOR;
 import static org.eclipse.jface.preference.JFacePreferences.INFORMATION_FOREGROUND_COLOR;
 
 public abstract class SonarLintWebView extends Composite implements Listener, IPropertyChangeListener {
 
+  protected static final String RULES_CONFIGURATION_LINK = "sonarlint://rules-configuration";
+  private static final String RULES_CONFIGURATION_ID = "org.sonarlint.eclipse.ui.properties.RulesConfigurationPage";
```
Comment (henryju) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintWebView.java:63]: Move this constant to `RulesConfigurationPage` class

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintRuleBrowser.java [hunk 11]
```diff
@@ -67,9 +78,52 @@ protected String body() {
         + "<img class=\"severityicon\" alt=\"" + severity + "\" src=\"data:image/gif;base64," + severityImg64 + "\">"
         + "<span>" + clean(severity) + "</span>"
         + "</div>"
-        + "<div class=\"rule-desc\">" + htmlDescription
-        + "</div>";
+        + "<div class=\"rule-desc\">" + htmlDescription + "</div>"
```
Comment (henryju) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/SonarLintRuleBrowser.java:81]: Try to not output the div when there is no parameter (in connected mode)

**Review by henryju — APPROVED** [review]
Minor feedback to handle then LGTM

**Pull request merged**